src/cider/nrepl/middleware/reload.clj

```diff
@@ -0,0 +1,58 @@
+(ns cider.nrepl.middleware.reload
+  (:require
+   [clj-reload.core :as reload]
+   [clojure.main :refer [repl-caught]]
+   [clojure.string :as str]
+   [haystack.analyzer :as analyzer]
+   [nrepl.middleware.interruptible-eval :refer [*msg*]]
+   [nrepl.middleware.print :as print]
+   [nrepl.misc :refer [response-for]]
+   [nrepl.transport :as transport]
+   [orchard.misc :as misc]))
+
+(defn- user-reload
+  "clj-reload.core/reload from the user project.
+  Must be configured via clj-reload.core/init before being called."
+  []
+  (some-> (symbol "clj-reload.core" "reload") ;; Don't use mrandorsenized version
+          resolve))
+
+(defn- init
+  "Initialize clj-reload with dirs.
+  Only used for test, but necessary because of mranderson."
+  [dirs]
+  (reload/init {:dirs dirs}))
+
+(defn respond
+  [{:keys [transport] :as msg} response]
+  (transport/send transport (response-for msg response)))
+
+(defn- refresh-reply
+  [{:keys [::print/print-fn transport session id] :as msg}]
+  (let [{:keys [exec]} (meta session)]
+    (exec id
+          (fn []
+            (try
+              ;; TODO: clear op
+              (let [reload (or (user-reload) reload/reload)]
+                (reload (cond-> {:log-fn (fn [& args]
+                                           (respond msg {:progress (str/join " " args)}))}
+                          (:all msg) (assoc :only :all)))
+                (respond msg {:status :ok}))
+              (catch Throwable error
+                (respond msg {:status :error
+                              ;; TODO assoc :file, :line info if available
+                              :error  (analyzer/analyze error print-fn)})
+                (binding [*msg* msg
+                          *err* (print/replying-PrintWriter :err msg {})]
+                  (repl-caught error)))))
+
+          (fn [] (respond msg {:status :done})))))
+
+(defn handle-reload [handler msg]
+  (case (:op msg)
+    "cider.clj-reload/reload" (refresh-reply msg)
+    "cider.clj-reload/reload-all" (refresh-reply (assoc msg :all true))
+    "cider.clj-reload/reload-clear" (refresh-reply (assoc msg :clear true))
```

Comment:
@vemv left support for the clear op here so it exists when called from CIDER, but it's still not functional as per the TODO in this file.